RecoLocalTracker/SiPixelClusterizer/plugins/alpaka/PixelClustering.h

```diff
@@ -0,0 +1,451 @@
+#include <cstdio>
+#include <type_traits>
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h"
+#include "DataFormats/SiPixelClusterSoA/interface/ClusteringConstants.h"
+#include "Geometry/CommonTopologies/interface/SimplePixelTopology.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h"
+
+// #define GPU_DEBUG
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  namespace pixelClustering {
+
+#ifdef GPU_DEBUG
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_STATIC_ACC_MEM_GLOBAL uint32_t gMaxHit = 0;
+#endif
+
+    namespace PixelStatus {
+      // Phase-1 pixel modules
+      constexpr uint32_t pixelSizeX = pixelTopology::Phase1::numRowsInModule;
+      constexpr uint32_t pixelSizeY = pixelTopology::Phase1::numColsInModule;
+
+      // Use 0x00, 0x01, 0x03 so each can be OR'ed on top of the previous ones
+      enum Status : uint32_t { kEmpty = 0x00, kFound = 0x01, kDuplicate = 0x03 };
+
+      constexpr uint32_t bits = 2;
+      constexpr uint32_t mask = (0x01 << bits) - 1;
+      constexpr uint32_t valuesPerWord = sizeof(uint32_t) * 8 / bits;
+      constexpr uint32_t size = pixelSizeX * pixelSizeY / valuesPerWord;
+
+      ALPAKA_FN_ACC ALPAKA_FN_INLINE constexpr static uint32_t getIndex(uint16_t x, uint16_t y) {
+        return (pixelSizeX * y + x) / valuesPerWord;
+      }
+
+      ALPAKA_FN_ACC ALPAKA_FN_INLINE constexpr static uint32_t getShift(uint16_t x, uint16_t y) {
+        return (x % valuesPerWord) * 2;
+      }
+
+      ALPAKA_FN_ACC ALPAKA_FN_INLINE constexpr static Status getStatus(uint32_t const* __restrict__ status,
+                                                                       uint16_t x,
+                                                                       uint16_t y) {
+        uint32_t index = getIndex(x, y);
+        uint32_t shift = getShift(x, y);
+        return Status{(status[index] >> shift) & mask};
+      }
+
+      ALPAKA_FN_ACC ALPAKA_FN_INLINE constexpr static bool isDuplicate(uint32_t const* __restrict__ status,
+                                                                       uint16_t x,
+                                                                       uint16_t y) {
+        return getStatus(status, x, y) == kDuplicate;
+      }
+
+      template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+      ALPAKA_FN_ACC ALPAKA_FN_INLINE constexpr static void promote(TAcc const& acc,
+                                                                   uint32_t* __restrict__ status,
+                                                                   const uint16_t x,
+                                                                   const uint16_t y) {
+        uint32_t index = getIndex(x, y);
+        uint32_t shift = getShift(x, y);
+        uint32_t old_word = status[index];
```

Comment:
I'm not sure I follow.
The whole point of the `atomicCas` is to make the change to `status[index]` only if it did not change since the read, _i.e._ if `expected` matches `old_word`.
Doesn't that guarantee that there is no race ?

Reply:
The possible data race would be between one thread doing this non-atomic read, while another thread is doing the `atomicCas()` operation on the same memory address.

Reply:
I'm not sure I follow.

CUDA does not have the concept of "atomic read". The fact that one thread is doing a non-atomic read while a different thread does an atomic CAS is normal.

One possible concern could be that the compiler caches a value instead of doing the read from global memory: to avoid this, the CUDA documentation suggests adding the `volatile` qualifier to the pointer type (yes, this is different from the C++ memory model).

What other kid of race can there be ?

Reply:
I'm indeed thinking in terms of C++ memory model, and beyond the CPU serial and CUDA backends we use today. CPU backends should be fine as long as Alpaka grid/block is mapped to a CPU thread (which is likely we'll do in any case, i.e. to not map Alpaka thread to CPU thread).

What about ROCm? (I'd assume AMD to mimic NVIDIA behavior here, but I don't know)

What about SYCL? Can it depend on the SYCL device type? (I guess we'd be mostly interested in the behavior on Intel GPUs, but who knows)

Ideally it would be nice if Alpaka documentation would include notes how the memory model behavior differs from standard C++.

Reply:
I assumes HIP/ROCm matches the behaviour of CUDA.

I think you are overestimating the quality of the documentation of Alpaka :-)

Reply:
> I think you are overestimating the quality of the documentation of Alpaka :-)

I know the reality, but want to make wishes anyway :)

Would it be worth of adding some comment here why this is not a data race in practice?

Reply:
If the current tests pass, I'd rather not touch this PR any more :-)

However, we should open an issue to collect all the "to do later" changes.

Reply:
Ok